File: src/main/java/org/example/eshop/service/AdminOrderService.java
package org.example.eshop.service;

import org.example.eshop.dto.*;
import org.example.eshop.entity.*;
import org.example.eshop.repository.OrderItemRepository;
import org.example.eshop.repository.OrderRepository;
import org.example.eshop.repository.ProductRepository;
import org.example.eshop.repository.VariantRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

@Service
@Transactional
public class AdminOrderService {

    private static final Logger logger = LoggerFactory.getLogger(AdminOrderService.class);

    private final OrderRepository orderRepository;
    private final OrderItemRepository orderItemRepository;
    private final VariantRepository variantRepository;
    private final ProductRepository productRepository;

    public AdminOrderService(OrderRepository orderRepository,
                             OrderItemRepository orderItemRepository,
                             VariantRepository variantRepository,
                             ProductRepository productRepository) {
        this.orderRepository = orderRepository;
        this.orderItemRepository = orderItemRepository;
        this.variantRepository = variantRepository;
        this.productRepository = productRepository;
    }

    public AdminOrderListResponse getAllOrders(int page, int size, String status, String paymentStatus, String fulfillmentStatus) {
        OrderStatus orderStatus = null;
        PaymentStatus paymentStatusEnum = null;
        FulfillmentStatus fulfillmentStatusEnum = null;

        if (status != null) {
            try { orderStatus = OrderStatus.valueOf(status.toUpperCase()); }
            catch (IllegalArgumentException e) { throw new IllegalArgumentException("Invalid order status: " + status); }
        }
        if (paymentStatus != null) {
            try { paymentStatusEnum = PaymentStatus.valueOf(paymentStatus.toUpperCase()); }
            catch (IllegalArgumentException e) { throw new IllegalArgumentException("Invalid payment status: " + paymentStatus); }
        }
        if (fulfillmentStatus != null) {
            try { fulfillmentStatusEnum = FulfillmentStatus.valueOf(fulfillmentStatus.toUpperCase()); }
            catch (IllegalArgumentException e) { throw new IllegalArgumentException("Invalid fulfillment status: " + fulfillmentStatus); }
        }

        Pageable pageRequest = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"));
        Page<Order> ordersPage = orderRepository.findAllWithFilters(orderStatus, paymentStatusEnum, fulfillmentStatusEnum, pageRequest);

        List<AdminOrderSummaryResponse> summaries = ordersPage.getContent().stream().map(order ->
                new AdminOrderSummaryResponse(
                        order.getId(),
                        order.getNumber(),
                        order.getEmail(),
                        order.getTotal(),
                        order.getStatus(),
                        order.getPaymentStatus(),
                        order.getFulfillmentStatus(),
                        order.getCreatedAt(),
                        order.getItems().size()
                )
        ).collect(Collectors.toList());

        return new AdminOrderListResponse(
                summaries,
                ordersPage.getTotalElements(),
                ordersPage.getTotalPages(),
                page,
                size
        );
    }

    public AdminOrderResponse getOrderById(Long id) {
        Order order = orderRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Order with id " + id + " not found"));
        return toAdminOrderResponse(order);
    }

    public AdminOrderResponse markOrderAsPaid(Long id) {
        Order order = orderRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Order with id " + id + " not found"));

        if (order.getPaymentStatus() == PaymentStatus.PAID) {
            throw new IllegalStateException("Order " + order.getNumber() + " is already marked as paid");
        }
        if (order.getStatus() == OrderStatus.CANCELLED) {
            throw new IllegalStateException("Cannot mark cancelled order " + order.getNumber() + " as paid");
        }

        order.setPaymentStatus(PaymentStatus.PAID);
        order.setStatus(OrderStatus.CONFIRMED);

        order.getItems().forEach(orderItem -> {
            Variant variant = variantRepository.findById(orderItem.getVariantId())
                    .orElseThrow(() -> new IllegalStateException("Variant " + orderItem.getVariantId() + " not found for order item " + orderItem.getId()));

            if (variant.getReservedQty() < orderItem.getQty()) {
                throw new IllegalStateException("Insufficient reserved stock for variant " + variant.getSku() + ". " +
                        "Required: " + orderItem.getQty() + ", Reserved: " + variant.getReservedQty());
            }

            int oldStock = variant.getStockQty();
            int oldReserved = variant.getReservedQty();
            variant.setStockQty(variant.getStockQty() - orderItem.getQty());
            variant.setReservedQty(variant.getReservedQty() - orderItem.getQty());
            variantRepository.save(variant);

            logger.info("Stock adjusted for variant {}: stockQty {} -> {}, reservedQty {} -> {}",
                    variant.getSku(), oldStock, variant.getStockQty(), oldReserved, variant.getReservedQty());
        });

        Order saved = orderRepository.save(order);

        logger.info("Payment Received - Order: {}, Items: {}, Total: {}, VAT: {}, Shipping: {}",
                order.getNumber(),
                order.getItems().stream().map(i -> i.getTitleSnapshot() + " (" + i.getQty() + ")").collect(Collectors.toList()),
                order.getTotal(), order.getTax(), order.getShipping());

        return toAdminOrderResponse(saved);
    }

    public AdminOrderResponse shipOrder(Long id, String trackingUrl) {
        Order order = orderRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Order with id " + id + " not found"));

        if (order.getPaymentStatus() != PaymentStatus.PAID) {
            throw new IllegalStateException("Cannot ship unpaid order " + order.getNumber());
        }
        if (order.getFulfillmentStatus() == FulfillmentStatus.FULFILLED) {
            throw new IllegalStateException("Order " + order.getNumber() + " is already shipped");
        }
        if (order.getStatus() == OrderStatus.CANCELLED) {
            throw new IllegalStateException("Cannot ship cancelled order " + order.getNumber());
        }

        order.setFulfillmentStatus(FulfillmentStatus.FULFILLED);
        order.setTrackingUrl(trackingUrl);

        Order saved = orderRepository.save(order);
        String trackingInfo = trackingUrl != null ? "Tracking: " + trackingUrl : "No tracking provided";
        logger.info("Order Shipped - Order: {}, {}", order.getNumber(), trackingInfo);
        return toAdminOrderResponse(saved);
    }

    public AdminOrderResponse cancelOrder(Long id) {
        Order order = orderRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Order with id " + id + " not found"));

        if (order.getStatus() == OrderStatus.CANCELLED) {
            throw new IllegalStateException("Order " + order.getNumber() + " is already cancelled");
        }
        if (order.getFulfillmentStatus() == FulfillmentStatus.FULFILLED) {
            throw new IllegalStateException("Cannot cancel a fulfilled order " + order.getNumber());
        }

        // Revert stock reservations if the order was previously reserved but not paid
        order.getItems().forEach(orderItem -> {
            Variant variant = variantRepository.findById(orderItem.getVariantId())
                    .orElseThrow(() -> new IllegalStateException("Variant " + orderItem.getVariantId() + " not found for order item " + orderItem.getId()));

            int oldReserved = variant.getReservedQty();
            variant.setReservedQty(Math.max(0, variant.getReservedQty() - orderItem.getQty()));
            variantRepository.save(variant);

            logger.info("Reservation reverted for variant {}: reservedQty {} -> {}",
                    variant.getSku(), oldReserved, variant.getReservedQty());
        });

        order.setStatus(OrderStatus.CANCELLED);
        Order saved = orderRepository.save(order);
        logger.info("Order Cancelled - Order: {}", order.getNumber());
        return toAdminOrderResponse(saved);
    }

    private AdminOrderResponse toAdminOrderResponse(Order order) {
        List<AdminOrderItemResponse> itemResponses = order.getItems().stream().map(orderItem -> {
            Variant variant = variantRepository.findById(orderItem.getVariantId()).orElse(null);
            Product product = (variant != null) ? productRepository.findById(variant.getProductId()).orElse(null) : null;
            return new AdminOrderItemResponse(
                    orderItem.getId(),
                    orderItem.getVariantId(),
                    orderItem.getTitleSnapshot(),
                    orderItem.getQty(),
                    orderItem.getPriceSnapshot(),
                    orderItem.getLineTotal(),
                    variant != null ? variant.getSku() : null,
                    product != null ? product.getTitle() : null
            );
        }).collect(Collectors.toList());

        return new AdminOrderResponse(
                order.getId(),
                order.getNumber(),
                order.getEmail(),
                order.getAddress(),
                order.getSubtotal(),
                order.getTax(),
                order.getShipping(),
                order.getTotal(),
                order.getStatus(),
                order.getPaymentStatus(),
                order.getFulfillmentStatus(),
                order.getTrackingUrl(),
                order.getCreatedAt(),
                order.getUpdatedAt(),
                itemResponses
        );
    }
}

File: src/main/kotlin/org/example/eshop/service/AdminOrderService.kt
package org.example.eshop.service

import org.example.eshop.dto.*
import org.example.eshop.entity.*
import org.example.eshop.repository.*
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.slf4j.LoggerFactory
import java.math.BigDecimal

@Service
@Transactional
class AdminOrderService(
    private val orderRepository: OrderRepository,
    private val orderItemRepository: OrderItemRepository,
    private val variantRepository: VariantRepository,
    private val productRepository: ProductRepository
) {
    
    private val logger = LoggerFactory.getLogger(AdminOrderService::class.java)

    fun getAllOrders(
        page: Int,
        size: Int,
        status: String?,
        paymentStatus: String?,
        fulfillmentStatus: String?
    ): AdminOrderListResponse {
        // Validate enum values if provided
        val orderStatus = status?.let { 
            try { OrderStatus.valueOf(it.uppercase()) } 
            catch (e: IllegalArgumentException) { 
                throw IllegalArgumentException("Invalid order status: $it") 
            }
        }
        val paymentStatusEnum = paymentStatus?.let { 
            try { PaymentStatus.valueOf(it.uppercase()) } 
            catch (e: IllegalArgumentException) { 
                throw IllegalArgumentException("Invalid payment status: $it") 
            }
        }
        val fulfillmentStatusEnum = fulfillmentStatus?.let { 
            try { FulfillmentStatus.valueOf(it.uppercase()) } 
            catch (e: IllegalArgumentException) { 
                throw IllegalArgumentException("Invalid fulfillment status: $it") 
            }
        }

        val pageRequest = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"))
        val ordersPage = orderRepository.findAllWithFilters(
            orderStatus, paymentStatusEnum, fulfillmentStatusEnum, pageRequest
        )

        val orderSummaries = ordersPage.content.map { order ->
            AdminOrderSummaryResponse(
                id = order.id,
                number = order.number,
                email = order.email,
                total = order.total,
                status = order.status,
                paymentStatus = order.paymentStatus,
                fulfillmentStatus = order.fulfillmentStatus,
                createdAt = order.createdAt,
                itemCount = order.items.size
            )
        }

        return AdminOrderListResponse(
            orders = orderSummaries,
            totalElements = ordersPage.totalElements,
            totalPages = ordersPage.totalPages,
            currentPage = page,
            pageSize = size
        )
    }

    fun getOrderById(id: Long): AdminOrderResponse {
        val order = orderRepository.findById(id).orElseThrow {
            NoSuchElementException("Order with id $id not found")
        }

        return toAdminOrderResponse(order)
    }

    fun markOrderAsPaid(id: Long): AdminOrderResponse {
        val order = orderRepository.findById(id).orElseThrow {
            NoSuchElementException("Order with id $id not found")
        }

        // Validate order state
        if (order.paymentStatus == PaymentStatus.PAID) {
            throw IllegalStateException("Order ${order.number} is already marked as paid")
        }
        if (order.status == OrderStatus.CANCELLED) {
            throw IllegalStateException("Cannot mark cancelled order ${order.number} as paid")
        }

        // Update payment status
        order.paymentStatus = PaymentStatus.PAID
        order.status = OrderStatus.CONFIRMED

        // Adjust stock quantities: stockQty -= qty, reservedQty -= qty
        order.items.forEach { orderItem ->
            val variant = variantRepository.findById(orderItem.variantId).orElseThrow {
                IllegalStateException("Variant ${orderItem.variantId} not found for order item ${orderItem.id}")
            }

            // Check if we have enough reserved stock
            if (variant.reservedQty < orderItem.qty) {
                throw IllegalStateException(
                    "Insufficient reserved stock for variant ${variant.sku}. " +
                    "Required: ${orderItem.qty}, Reserved: ${variant.reservedQty}"
                )
            }

            // Adjust stock quantities
            val updatedVariant = variant.copy(
                stockQty = variant.stockQty - orderItem.qty,
                reservedQty = variant.reservedQty - orderItem.qty
            )
            variantRepository.save(updatedVariant)

            logger.info("Stock adjusted for variant ${variant.sku}: " +
                "stockQty ${variant.stockQty} -> ${updatedVariant.stockQty}, " +
                "reservedQty ${variant.reservedQty} -> ${updatedVariant.reservedQty}")
        }

        val savedOrder = orderRepository.save(order)

        // Log payment received event
        logger.info("Payment Received - Order: ${order.number}, " +
            "Items: ${order.items.map { "${it.titleSnapshot} (${it.qty})" }}, " +
            "Total: ${order.total}, VAT: ${order.tax}, Shipping: ${order.shipping}")

        return toAdminOrderResponse(savedOrder)
    }

    fun shipOrder(id: Long, trackingUrl: String?): AdminOrderResponse {
        val order = orderRepository.findById(id).orElseThrow {
            NoSuchElementException("Order with id $id not found")
        }

        // Validate order state
        if (order.paymentStatus != PaymentStatus.PAID) {
            throw IllegalStateException("Cannot ship unpaid order ${order.number}")
        }
        if (order.fulfillmentStatus == FulfillmentStatus.FULFILLED) {
            throw IllegalStateException("Order ${order.number} is already shipped")
        }
        if (order.status == OrderStatus.CANCELLED) {
            throw IllegalStateException("Cannot ship cancelled order ${order.number}")
        }

        // Update fulfillment status
        order.fulfillmentStatus = FulfillmentStatus.FULFILLED
        order.trackingUrl = trackingUrl

        val savedOrder = orderRepository.save(order)

        // Log order shipped event
        val trackingInfo = trackingUrl?.let { "Tracking: $it" } ?: "No tracking provided"
        logger.info("Order Shipped - Order: ${order.number}, $trackingInfo")

        return toAdminOrderResponse(savedOrder)
    }

    fun cancelOrder(id: Long): AdminOrderResponse {
        val order = orderRepository.findById(id).orElseThrow {
            NoSuchElementException("Order with id $id not found")
        }

        // Validate order state
        if (order.status == OrderStatus.CANCELLED) {
            throw IllegalStateException("Order ${order.number} is already cancelled")
        }
        if (order.paymentStatus == PaymentStatus.PAID) {
            throw IllegalStateException("Cannot cancel paid order ${order.number}. Refund required.")
        }
        if (order.fulfillmentStatus == FulfillmentStatus.FULFILLED) {
            throw IllegalStateException("Cannot cancel shipped order ${order.number}")
        }

        // Update order status
        order.status = OrderStatus.CANCELLED

        // Release reservations: reservedQty -= qty
        order.items.forEach { orderItem ->
            val variant = variantRepository.findById(orderItem.variantId).orElseThrow {
                IllegalStateException("Variant ${orderItem.variantId} not found for order item ${orderItem.id}")
            }

            // Check if we have enough reserved stock to release
            if (variant.reservedQty < orderItem.qty) {
                logger.warn("Inconsistent reserved stock for variant ${variant.sku}. " +
                    "Expected: ${orderItem.qty}, Reserved: ${variant.reservedQty}")
            }

            // Release reservation
            val updatedVariant = variant.copy(
                reservedQty = maxOf(0, variant.reservedQty - orderItem.qty)
            )
            variantRepository.save(updatedVariant)

            logger.info("Reservation released for variant ${variant.sku}: " +
                "reservedQty ${variant.reservedQty} -> ${updatedVariant.reservedQty}")
        }

        val savedOrder = orderRepository.save(order)

        // Log order cancellation
        logger.info("Order Cancelled - Order: ${order.number}")

        return toAdminOrderResponse(savedOrder)
    }

    private fun toAdminOrderResponse(order: Order): AdminOrderResponse {
        val orderItems = order.items.map { orderItem ->
            val variant = variantRepository.findById(orderItem.variantId).orElse(null)
            val product = variant?.let { productRepository.findById(it.productId).orElse(null) }

            AdminOrderItemResponse(
                id = orderItem.id,
                variantId = orderItem.variantId,
                titleSnapshot = orderItem.titleSnapshot,
                qty = orderItem.qty,
                priceSnapshot = orderItem.priceSnapshot,
                lineTotal = orderItem.priceSnapshot.multiply(BigDecimal(orderItem.qty)),
                sku = variant?.sku,
                productTitle = product?.title
            )
        }

        return AdminOrderResponse(
            id = order.id,
            number = order.number,
            email = order.email,
            address = order.address,
            subtotal = order.subtotal,
            tax = order.tax,
            shipping = order.shipping,
            total = order.total,
            status = order.status,
            paymentStatus = order.paymentStatus,
            fulfillmentStatus = order.fulfillmentStatus,
            trackingUrl = order.trackingUrl,
            createdAt = order.createdAt,
            updatedAt = order.updatedAt,
            items = orderItems
        )
    }
}